### PR TITLE
fix: harden ServiceMonitor (port detection, selector, interval)

### PR DESCRIPTION
**Pull request by jan-ludvik**
## Summary

Hardens the `ServiceMonitor` constructor in `kubernetes/kube.libsonnet`: fixes a latent metrics-port detection bug, fails loudly on ambiguous port config, narrows the selector, and makes the scrape interval a parameter. Metric-shaping defaults (labels, honorLabels) are left at their permissive values on purpose — see below.

## Usage

Minimal (stencil service with a `metrics` port):

```jsonnet
servicemonitor: ok.ServiceMonitor(app.name, app.namespace) {
  target_service:: $.service,
},
```

With overrides (all optional):

```jsonnet
servicemonitor: ok.ServiceMonitor(app.name, app.namespace, interval='60s') {
  target_service:: $.service,
  spec+: {
    endpoints_+:: { 'http-prom': {
      interval: '15s',                                                              // per-endpoint tuning, keyed by targetPort
      metricRelabelings: [{ sourceLabels: ['__name__'], regex: 'go_gc_.*', action: 'drop' }],  // drop noisy series
    } },
    targetLabels_:: ['app', 'repo'],                                                // restrict copied labels (default: all)
  },
},
```

## Changes

| # | Change | Why |
|---|--------|-----|
| 1 | Port detection uses `std.member`, not `std.setMember` | `setMember` binary-searches an assumed-sorted set; on the unsorted `[p.name, p.targetPort]` it misdetected a port with `targetPort: metrics` whose `name` sorts after `metrics`. |
| 2 | No `metrics` port → sole-port fallback, else **error** | The old code fell back to `spec.ports[0]` (alphabetically first — could be gRPC). Single-port services still work; multi-port-no-metrics now fails loudly. Override with `spec.endpoints_`. |
| 3 | `selector.matchLabels` → `{ name: <svc> }` (via `selectorLabels_::`) | Old default matched the full label set, silently breaking if any label changed. |
| 4 | Scrape `interval` is a constructor arg, default `30s` | Was hardcoded `1m`. Still overridable per-endpoint via `endpoints_`. |
| 5 | Removed the `prometheus.io/scrape` label | The Target Allocator's `serviceMonitorSelector` is open (match-all), so the label selected nothing and only confused readers. |

**Left permissive on purpose:** `targetLabels` copies **all** Service labels onto the series (low-cardinality dimensions like `repo`/`bento`/`reporting_team` that downstream consumers need — this is the only place to surface them), and `honorLabels` stays `true`. Label renames (e.g. `reporting_team → team`) are handled **centrally in the collector pipeline**, not per-ServiceMonitor. Per-endpoint `metricRelabelings` pass through via `endpoints_`.

## Compatibility

Stencil-golang always emits a `metrics`-named Service port, so default detection works for every current/future stencil service; the two live callers (`ds-mlflow-project-proxy`, `tokenauthenticator`) are single-port and covered by the fallback. Verified both still render.

## Testing

`tests/kube.jsonnet` covers port discovery (the `std.member` edge case), single-port fallback, `honorLabels == true`, `interval == 30s` (+ override), copy-all `targetLabels`, `{ name }` selector, absence of any scrape marker label, and per-endpoint `metricRelabelings` pass-through. `make test` passes; no snapshot drift.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: kubernetes/kube.libsonnet [hunk 1]
```diff
@@ -793,32 +821,67 @@ local environment = std.extVar('environment');
...
+      function(p) std.member([p.name, p.targetPort], 'metrics'),
+      ports,
+    ),
+    local default_port =
+      if std.length(matched_ports) > 0 then matched_ports[0]
+      // No 'metrics' port: fall back to the sole port when unambiguous
+      // (common for apps serving /metrics on their only port).
+      else if std.length(ports) == 1 then ports[0]
+      // Multiple ports and none is 'metrics' -> refuse to guess (objectFields
+      // sorting could pick e.g. the gRPC port). Set spec.endpoints_ explicitly.
```
Comment (AndrewWinterman) [inline kubernetes/kube.libsonnet:841]: let's default to the stencil-golang standard in this case, 8000: https://github.com/getoutreach/stencil-golang/blob/d24f1ee5384a4ae6e57cb31b41b2edf334e7972b/templates/deployments/appname/app.jsonnet.tpl#L115-L117

Comment (AndrewWinterman) [reply kubernetes/kube.libsonnet:841]: actually i take back-- erroring at render time is better

File: kubernetes/kube.libsonnet [hunk 2]
```diff
@@ -793,32 +821,67 @@ local environment = std.extVar('environment');
...
+    // Central metric_relabel_configs applied to every endpoint, appended AFTER
+    // any caller-supplied rules so platform-wide cardinality/cost guardrails
+    // cannot be dropped by a team's endpoints_ entry. Populate cluster-wide
+    // drop/keep rules here.
+    centralMetricRelabelings:: [],
+
+    // Marker label the OTel Target Allocator's serviceMonitorSelector keys off.
+    // Keep in sync with the collector's
+    // targetAllocator.prometheusCR.serviceMonitorSelector.
+    scrapeLabels_:: { 'monitoring.outreach.io/otel-scrape': 'true' },
```
Comment (AndrewWinterman) [inline kubernetes/kube.libsonnet]: we should scrape the servicemonitor default. Better to conform to the standard.

On clickhouse especially we don't particularly need to filter metrics out.

File: kubernetes/kube.libsonnet [hunk 3]
```diff
@@ -793,32 +821,67 @@ local environment = std.extVar('environment');
...
+
+    metadata+: { labels+: this.scrapeLabels_ },
     spec: {
+      // Labels copied from the target Service onto every scraped series. Keep
+      // this minimal -- each entry multiplies label cardinality (and storage)
```
Comment (AndrewWinterman) [inline kubernetes/kube.libsonnet]: keep it maximal actually. Labels are unlikely to lead to cardinality explosion, unless they vary by pod which is unusual, and in clickhouse high cardinality is fine anyhow.

File: kubernetes/kube.libsonnet [hunk 4]
```diff
@@ -793,32 +821,67 @@ local environment = std.extVar('environment');
...
       // this is just map-based sugar around self.endpoints
       endpoints_:: { [default_port.targetPort]: {} },
       // override this to explicitly adhere to the operator's API
       // and ignore all of the above, which is simply sugar
       endpoints: [
-        { honorLabels: true, interval: '1m', targetPort: p }
-        + this.spec.endpoints_[p]
+        // honorLabels defaults to false: app-exposed labels must not clobber
+        // topology labels (job/namespace/instance) or the jobLabel below.
+        local base = { honorLabels: false, interval: '1m', targetPort: p } + this.spec.endpoints_[p];
```
Comment (AndrewWinterman) [inline kubernetes/kube.libsonnet]: Better to default `honorLabels` to true. We also need to discuss collection interval and when we want to know increase it period. We should evaluate the cost of matching datadog's collection interval (10 seconds)

**Review by AndrewWinterman — APPROVED** [review]
Does the servicemonitor scrape pods or kubernetes service. IF it scrapes the service, how does it aggregate metrics from pods? Simply report whichever pod it found?

File: kubernetes/kube.libsonnet (hunk 2 above)
Comment (jan-ludvik) [reply kubernetes/kube.libsonnet]: Actually this doesn't do anything. TA has a selector which is now completely open - everything will be scraped. We can add the label but maybe we should just remove this so it doesn't confuse anyone

File: kubernetes/kube.libsonnet (hunk 4 above)
Comment (jan-ludvik) [reply kubernetes/kube.libsonnet]: done

Comment (jan-ludvik) [reply kubernetes/kube.libsonnet]: Yes the interval is a question - I don't have a preference really and would go with something like `30s` if I had to pick - but I don't want to pick. It should be definitely overridable which I will change

File: kubernetes/kube.libsonnet (hunk 3 above)
Comment (jan-ludvik) [reply kubernetes/kube.libsonnet]: done

File: kubernetes/kube.libsonnet (hunk 2 above)
Comment (jan-ludvik) [reply kubernetes/kube.libsonnet]: removed

File: kubernetes/kube.libsonnet (hunk 4 above)
Comment (jan-ludvik) [reply kubernetes/kube.libsonnet]: I will merge with 30s which seems sensible and we can make tiny update if you think that's too long

**Review by ckopparthi — APPROVED** [review]
